core/src/main/scala/org/bitcoins/core/psbt/PSBT.scala

```diff
@@ -58,11 +67,218 @@ case class PSBT(
       vec.head.version
     }
   }
+
+  /**
+    * Adds tx to the indexed InputPSBTMap to either the NonWitnessOrUnknownUTXO
+    * or WitnessUTXO field depending on the tx and available information in the PSBT
+    * @param tx Transaction to add to PSBT
+    * @param index index of the InputPSBTMap to add tx to
+    * @return PSBT with added tx
+    */
+  def addTransactionToInput(tx: Transaction, index: Int): PSBT = {
+    require(
+      index < inputMaps.size,
+      s"i must be less than the number of input maps present in the psbt, $index >= ${inputMaps.size}")
+    require(!inputMaps(index).isFinalized,
+            s"Cannot update an InputPSBTMap that is finalized, index: $index")
+
+    var elements = inputMaps(index).elements
+    val txIn = transaction.inputs(index)
+
+    if (txIn.previousOutput.vout.toInt < tx.outputs.size) {
+      val out = tx.outputs(txIn.previousOutput.vout.toInt)
+
+      val outIsWitnessScript =
+        WitnessScriptPubKey.isWitnessScriptPubKey(out.scriptPubKey.asm)
+      val hasWitScript =
+        inputMaps(index)
+          .getRecords[InputPSBTRecord.WitnessScript](WitnessScriptKeyId)
+          .size == 1
+
+      if (outIsWitnessScript || hasWitScript) {
+        elements = elements.filterNot(_.key.head == WitnessUTXOKeyId.byte) :+ WitnessUTXO(
+          out)
+      } else {
+        elements = elements.filterNot(_.key.head == NonWitnessUTXOKeyId.byte) :+ NonWitnessOrUnknownUTXO(
+          tx)
+      }
+    }
```

Comment:
Rather than using a `var` for elements, I think it would be more idiomatic to do `val elements = if ...` and then just drop the `elements =` on the ends, and rather than elements call `inputMaps(index).elements`, and also add an else case which just returns `inputMaps(index).elements`. Or something of this nature

Reply:
you can also give `inputMaps(index).elements` a `val` if it comes up enough